evm-contracts/src/v0.6/VRFCoordinator.sol

```diff
@@ -110,70 +116,130 @@ contract VRFCoordinator is VRF, VRFRequestIDBase {
    *
-   * @dev This is the main entrypoint for chainlink. If you change this, you
-   * @dev should also change the solidityABISstring in solidity_proof.go.
+   * @dev The structure of _proof corresponds to vrf.MarshaledOnChainResponse,
+   * @dev in the node source code. I.e., it is a vrf.MarshaledProof with the seed
+   * @dev replaced by the preseed, followed by the hash of the requesting block.
    */
-  function fulfillRandomnessRequest(bytes memory _proof) public returns (bool) {
-    // TODO(alx): Replace the public key in the above proof with an argument
-    // specifying the keyHash. Splice the key in here before sending it to
-    // VRF.sol. Should be able to save about 2,000 gas that way.
-    // https://www.pivotaltracker.com/story/show/170828567
-    //
-    // TODO(alx): Move this parsing into VRF.sol, where the bytes layout is recorded.
-    // https://www.pivotaltracker.com/story/show/170828697
-    uint256[2] memory publicKey;
-    uint256 seed;
-    // solhint-disable-next-line no-inline-assembly
-    assembly { // Extract the public key and seed from proof
-      publicKey := add(_proof, 0x20) // Skip length word for first 64 bytes
-      seed := mload(add(_proof, 0xe0)) // Seed is 7th word in proof, plus word for length
-    }
-    bytes32 currentKeyHash = hashOfKey(publicKey);
-    bytes32 requestId = makeRequestId(currentKeyHash, seed);
-    Callback memory callback = callbacks[requestId];
-    require(callback.callbackContract != address(0), "no corresponding request");
-    uint256 randomness = VRF.randomValueFromVRFProof(_proof); // Reverts on failure
+  function fulfillRandomnessRequest(bytes memory _proof) public {
+    (bytes32 currentKeyHash, Callback memory callback, bytes32 requestId,
+     uint256 randomness) = getRandomnessFromProof(_proof);
+
+    // Pay oracle
     address oadd = serviceAgreements[currentKeyHash].vRFOracle;
     withdrawableTokens[oadd] = withdrawableTokens[oadd].add(callback.randomnessFee);
+
+    // Forget request. Must precede callback (prevents reentrancy)
+    delete callbacks[requestId]; 
+    callBackWithRandomness(requestId, randomness, callback.callbackContract);
+  }
+
+  function callBackWithRandomness(bytes32 requestId, uint256 randomness,
+    address consumerContract) internal {
     // Dummy variable; allows access to method selector in next line. See
     // https://github.com/ethereum/solidity/issues/3506#issuecomment-553727797
     VRFConsumerBase v;
     bytes memory resp = abi.encodeWithSelector(
       v.fulfillRandomness.selector, requestId, randomness);
+    // The bound b here comes from https://eips.ethereum.org/EIPS/eip-150. The
+    // actual gas available to the consuming contract will be b-floor(b/64).
+    // This is chosen to leave the consuming contract ~200k gas, after the cost
+    // of the call itself.
+    uint256 b = 206000;
+    require(gasleft() >= b, "not enough gas for consumer");
```

Comment:
@coventry and I wanted to get your take on this, @se3000

Reply:
@kaleofduty , @se3000 , are we good with this?